### PR TITLE
Spike: Codecov - part 1

### PR DESCRIPTION
**Pull request by mejdoo**
**Codecov provides highly integrated tools to group, merge, archive and compare coverage reports.**

We'd like to integrate Codecov to generate coverage report in our PR's. This could be done by  following those steps:

- Authorise Codecov GitHub app to access to **wire-android** repository : **DONE**
- Generate Jacoco **xml** report for code coverage  : **Goal of this PR**
- Send reports to Codecov using Jenkins  : **TO DO**

## What's new in this PR?

I created a gradle task `jacocoTestReport`to generate code coverage xml reports for **app module**.

https://wearezeta.atlassian.net/browse/AN-6766
#### APK
[Download build #1727](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1727/artifact/build/artifact/wire-dev-PR2733-1727.apk)

### CODE REVIEW

File: app/jacoco/coverage.gradle [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+task jacocoTestReport(type: JacocoReport) {
+
+    reports {
+        xml.enabled = true
+        html.enabled = true
+    }
+
+    def fileFilter = ['**/R.class', '**/R$*.class', '**/BuildConfig.*', '**/Manifest*.*', '**/*Test*.*', 'android/**/*.*']
+    def kotlinTree = fileTree(dir: "$project.buildDir/tmp/kotlin-classes/devDebug/com/waz/zclient", excludes: fileFilter)
+    def javaScalaTree =fileTree(dir: "$project.buildDir/intermediates/javac/devDebug/compileDevDebugJavaWithJavac/classes/com/waz", excludes: fileFilter)
```
Comment (BradleyWire) [inline app/jacoco/coverage.gradle]: formatting here

File: app/jacoco/coverage.gradle [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+apply plugin: 'jacoco'
+
+task jacocoTestReport(type: JacocoReport) {
+
+    reports {
+        xml.enabled = true
+        html.enabled = true
+    }
+
+    def fileFilter = ['**/R.class', '**/R$*.class', '**/BuildConfig.*', '**/Manifest*.*', '**/*Test*.*', 'android/**/*.*']
```
Comment (BradleyWire) [inline app/jacoco/coverage.gradle]: Do we want to exclude anything else that we don't test from our own project? I'm thinking maybe classes ending with Module for Koin as one example. This seems like the standard filter for Jacoco, but, i know there are probably stuff unique to our project.

Comment (mejdoo) [reply app/jacoco/coverage.gradle]: I don't think we have to exclude Module files them. We want to know how much code we can test anyway. For example, we don't need to implement tests for fragments and activities as well but we don't have to to exclude them. I'm thinking also about integrating the task for **storage** and **zmessaging** modules in **part 2** to have a global percentage of the unit tests.

**Comment by mejdoo** [comment]
build again

**Pull request merged**